src/crimson/osd/object_context.h

```diff
@@ -121,13 +121,23 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
   bool recovery_read_marker = false;
 
   template <typename Lock, typename Func>
-  auto _with_lock(Lock&& lock, Func&& func) {
+  auto _with_lock(Lock& lock, Func&& func) {
     Ref obc = this;
-    return lock.lock().then([&lock, func = std::forward<Func>(func), obc]() mutable {
-      return seastar::futurize_invoke(func).finally([&lock, obc] {
-	lock.unlock();
+    auto maybe_fut = lock.lock();
+    if (maybe_fut) {
+      return (*maybe_fut).then(
+        [&lock, func = std::forward<Func>(func), obc]() mutable {
+        return seastar::futurize_invoke(func
+        ).finally([&lock, obc] {
+          lock.unlock();
+        });
       });
-    });
+    } else {
+      return seastar::futurize_invoke(func
+      ).finally([&lock, obc] {
+        lock.unlock();
+      });
+    }
```

Comment:
the commit message is missing the problem description,  the root cause analysis, and the summary of this change.  could you please elaborate a little bit on the root cause of

> tri_mutex::promote_from_read(): Assertion `readers == 1' failed.

and how this change addresses this problem?

Reply:
My understanding see https://github.com/ceph/ceph/pull/55488#issuecomment-2009065584 that caused `writers == 1` failure.

I think the issue is similar that `func` (to promote and load-obc) isn't called immediately upon the 1st locked. The returned future allows the 2nd reader enter and lock (causing readers > 1) even the future is available, before `func` is called.

Making lock() to return nullopt instead of a ready future in this change makes the 1st locking and calling `func` atomic.

Note that we have fixed several issues related to out-of-order and ready future.

Reply:
Added more information in the commit message.

Reply:
I don't see how this actually helps.  The bulk of the changes here allow for the various lock invocations to *dynamically* return std::nullopt allowing us to take the else branch of the if statement above

```
      return seastar::futurize_invoke(func
      ).finally([&lock, obc] {
        lock.unlock();
      });
```

That only solves the atomicity problem it if actually happens to return std::nullopt, though.  For this to be an actual solution, we'd have to know statically that it really will.  In that case, why allow the function to return a non-ready future at all?

I think this is essentially symmetric to the other similar PR that attempted to return a std::optional<future<>> -- the right solution is to identify why we are confident that a particular pathway doesn't need to block and directly call a variant that will never block.

Reply:
I see, if my understanding is correct above, we should be able to convert read/write promotion to a non-block version, will try.

BTW, the test with the current change looks passing, see https://github.com/ceph/ceph/pull/56912#issuecomment-2060278568

Reply:
The original promotion is already synchronous, so spcialize `_with_promotion()` to make sure the function is called atomically.

Proceed to tests.